src/nfa/interleaved.rs

```diff
@@ -0,0 +1,864 @@
+    pub fn ascii_case_insensitive(&mut self, yes: bool) -> &mut Builder {
+        self.noncontiguous.ascii_case_insensitive(yes);
+        self
+    }
+
+    /// Enable heuristic prefilter optimizations.
+    ///
+    /// This only applies when using [`Builder::build`] and not
+    /// [`Builder::build_from_noncontiguous`].
+    ///
+    /// See
+    /// [`AhoCorasickBuilder::prefilter`](crate::AhoCorasickBuilder::prefilter)
+    /// for more documentation and examples.
+    pub fn prefilter(&mut self, yes: bool) -> &mut Builder {
+        self.noncontiguous.prefilter(yes);
+        self
+    }
+
+    /// Whether to attempt to shrink the size of the automaton's alphabet or
+    /// not.
+    ///
+    /// This should never be enabled unless you're debugging an automaton.
+    /// Namely, disabling byte classes makes transitions easier to reason
+    /// about, since they use the actual bytes instead of equivalence classes.
+    /// Disabling this confers no performance benefit at search time.
+    ///
+    /// See
+    /// [`AhoCorasickBuilder::byte_classes`](crate::AhoCorasickBuilder::byte_classes)
+    /// for more documentation and examples.
+    pub fn byte_classes(&mut self, yes: bool) -> &mut Builder {
+        self.byte_classes = yes;
+        self
+    }
+}
+
+/// Result of interleaving states
+struct Interleave {
+    /// Mapping from the old state ids to the new ones.
+    index_to_state_id: Vec<StateID>,
+
+    /// Largest new state id. Useful to allocate the right size
+    /// for the final table.
+    max_state_id: StateID,
+
+    /// Indicates which indexes in the table are free.
+    available_indexes: AvailableIndexes,
+
+    /// First index to search at for a given offset.
+    /// That is, if needing to add in the table a transition with a given offset, the
+    /// search should start at first_available[offset] or later.
+    /// This is only used to speed up the algorithm, which could otherwise take
+    /// quite long when the number of states grows.
+    offset_to_search_index: [usize; 257],
+
+    /// Max state id for matching state.
+    max_matching_state_id: StateID,
+
+    /// Max state id for starting state.
+    max_starting_state_id: StateID,
+}
+
+/// Interleave states to reduce the size of the mapping and improve
+/// cache locality.
+///
+/// See documentation on [`NFA`] for more details.
+fn interleave(
+    nnfa: &noncontiguous::NFA,
+    byte_classes: &ByteClasses,
+) -> Result<Interleave, BuildError> {
+    let mut interleave = Interleave {
+        index_to_state_id: vec![NFA::DEAD; nnfa.states().len()],
+        max_state_id: NFA::DEAD,
+        available_indexes: AvailableIndexes::default(),
+        offset_to_search_index: [0_usize; 257],
+        max_matching_state_id: StateID::ZERO,
+        max_starting_state_id: StateID::ZERO,
+    };
+
+    let states = nnfa.states();
+    let mut states_to_visit = VecDeque::new();
+    let mut visited = HashSet::with_capacity(nnfa.states().len());
```

Comment:
Using `ahash` crate instead of the default hash tends to speed up hashing decently, when I've measured.

Reply:
Aye. But just to be clear, I do not really want to add dependencies. `aho-corasick` is a core part of the `regex` crate, and I try to keep its dependency tree as slim as I possibly can. It used to have more dependencies in the past, and they generally wound up being a headache.

Another idea here is to just inline a simple implementation of FNV. Assuming we don't need a DoS resistant hashmap.

(I haven't reviewed the code yet, but it looks like this is only construction time stuff. If so, then pulling in another crate to get a small speedup is definitely not worth it.)

Reply:
This is simply a "old state id" => bool mapping, so using a `vec![false; nnfa.states().len()]` instead should work just as well. I'll test it this evening